### PR TITLE
Version 0.9: help, bug fixes

### LINKED ISSUE #2: v0.75c: segmentation fault calling verify.c
I was too hasty uploading my v0.75c without more testing, as mode 3 `-upr` Suyama tests calls `verify.c` without preparation of `gwdata`, and predictably falls over:
```
cxc@192-168-1-10 cofact % ./quick_historical.sh: line 143: 20190 Segmentation fault: 11  cofact -ciov F17.proof -sep 17 31065037602817
./quick_historical.sh: line 146: 20191 Segmentation fault: 11  cofact -u F17.proof -sep 17 31065037602817 7751061099802522589358967058392886922693580423169
./quick_historical.sh: line 150: 20193 Segmentation fault: 11  cofact -u F18.proof -sep 18 13631489
./quick_historical.sh: line 153: 20194 Segmentation fault: 11  cofact -u F18.proof -sep 18 13631489 81274690703860512587777
./quick_historical.sh: line 157: 20195 Segmentation fault: 11  cofact -mu F19.proof -sep 19 70525124609 646730219521
./quick_historical.sh: line 160: 20196 Segmentation fault: 11  cofact -u F19.proof -sep 19 70525124609 646730219521 37590055514133754286524446080499713
```
I have a few things in mind to resolve this.

### PR DESCRIPTION
**Pull request by xanthe-cat**
No version change (remains at 0.75c)
- Revised help
- Proof verification only for Mersenne numbers
- Fixes #2, seg faults calling verify.c

### CODE REVIEW

**Comment by xanthe-cat** [comment]
This pull request should only be applied to the `cxc` branch. (Is that possible?)

This pull fixes #2 by ruling out the problem case, as it seems `verify.c` needs further modification to be able to verify a VDF proof of a Fermat number.

**Comment by gary641** [comment]
I reviewed the README.md file and overall it looks great. Here are a few comments:

- In the Mathematics section, you say a Mersenne number must have a prime exponent. According to https://mathworld.wolfram.com/MersenneNumber.html the exponent only needs to be an integer. A prime exponent is a necessary but not sufficient condition for the Mersenne number to be prime.

- In the Suyama cofactor test section, you say $A ≡ b^{F−1} (mod F)$. Since this is just the assignment of the result of the RHS calculation to A, would = be better than ≡ ?  I think of ≡ (congruent) as being used when the LHS and RHS are being compared for equivalence mod F. Assignments like this occur in several other places in your Readme. In some places, both would seem appropriate to use, such as “Here, $R = b^{M−1} ≡ 1 (mod M)$ is used to establish probable primality” or maybe for clarity “Here, if $R = b^{M−1} (mod M)$, $R ≡ 1 (mod M)$ is used to establish probable primality”.

- One minor thing: sometimes your Readme refers to George’s program as mprime, sometimes Prime95, and once Prime95/mprime.   In my Readme, for consistency I usually called it mprime, with mprime / Prime95 used at the first mention (to establish their identity) and in a few more places (to refresh the concept of identity).

**Comment by xanthe-cat** [comment]
Thanks Gary, (and apologies for replying to this somewhat asynchronously after I’d looked at a whole lot of other things!).

On the first point, as to what is a Mersenne: I am _sort of_ following the Cunningham book, which would perhaps define things something like this:
• the numbers 2<sup>_x_</sup>+1 where _x_ is a positive integer, form the **2,+** sequence.
• the numbers 2<sup>_x_</sup>+1 where _x_ is an integral power of 2, are the subset of **2,+** known as the Fermat numbers.
And similarly,
• the numbers 2<sup>_x_</sup>–1 where _x_ is a positive integer, form the **2,–** sequence.
• the numbers 2<sup>_x_</sup>–1 where _x_ is a prime, are the subset of **2,–** known as the Mersenne numbers.
I’m aware that’s at variance with the definition of Eric at Wolfram, but I’m happy with this, largely because Fermat seems to have been the first person to work out that all the non-Fermat 2,+ numbers and all the non-Mersenne 2,– numbers had to be composite; he had come up with the 2_kp_+1 factor insight and utilised it to speedily discover Mersenne factors while erroneously nearly believing the Fermat numbers were likely to all be prime.

Before and after Fermat plenty of people claimed perfect numbers that weren’t perfect (though Cataldi, about a half century before Fermat, had clearly twigged that the exponent in 2<sup>_x_</sup>–1 had to be prime), but Fermat and Mersenne were contemporaries at the same time when Fermat put this onto a firmer ground. The people centuries earlier than Fermat clearly weren’t talking about “Mersenne numbers” in relation to perfect numbers – and since 1640 we’ve had this better understanding about which Mersenne numbers can and can’t possibly be prime.

That’s the historical–mathematical justification; and the weaker computational justification, is that cofact 0.9 as programmed disallows all composite exponents, and any prime exponents outside the range of [3,2<sup>30</sup>). If I permitted composite exponents on the 2,– side there would be many more bug reports!

On the second point, again this seems to be a fine point where what the mathematics says and what we do computationally are subtly different. If computing resources of memory and speed were no issue we could simply evaluate the RHS without ever performing a modular reduction, and then just do a division at the end; but obviously that’s completely impractical, so instead we are performing a modular division as a part of every squaring. So when I look at something like $A \equiv b^{F-1}$ (mod $F$) I agree that computationally the LHS $A$ is being assigned (equality) the least positive value modulo $F$, but mathematically the RHS $b^{F-1}$ is some colossal number beyond practical computation which is nevertheless congruent to $A$ given the same modulus… I’m not totally convinced by my answer here. But I acknowledge your point and will look at it again and again until I satisfy myself or find myself unsatisfied.

On Prime95/mprime, yes almost all of the time I mean mprime (the command-line program) and not Prime95 (the GUI application), except for where I am introducing them together in general. This inconsistency is almost certainly from the documentation being redrafted on multiple occasions, and I will attempt to squash it as soon as the next full review.

**Comment by gary641** [comment]
On the first point: Interesting!  I guess this is another case of the math community not all being in agreement. Wikipedia implies this, saying "Numbers of the form M<sub>n</sub> = 2<sup>n</sup> − 1 without the primality requirement may be called Mersenne numbers. Sometimes, however, Mersenne numbers are defined to have the additional requirement that n be prime." 

On the second point, I don't think = vs $\equiv$ implies anything about how the RHS is calculated. Using = says that, however the RHS is calculated, the result is assigned to A. Using $\equiv$ (as I understand it) says that after both the LHS and RHS are reduced using the modulo operation, their values are equal. I think an assignment is what the text is trying to convey, not a modulo comparison.

**Comment by xanthe-cat** [comment]
I’ve added the GPU licence v3 and resolved the Prime95/mprime ambiguity mostly in favour of mprime (where I do mention Prime95 it’s usually in reference to the source code distribution).

I gather it’s actually very easy to add the LICENSE file to your branch, as there’s a little icon that looks like a set of balance scales, which when you click on it gives a set of possible licences.

For the time being I’m going to defer making a decision on the = vs ≡ dilemma as I do want to be consistent, but if I change it in one place I suspect I will have to change it elsewhere in a lot more places.

If you agree with using the GPU licence v3 then I think that’s the last thing needed for me to merge this PR with the cxc branch.

**Comment by xanthe-cat** [comment]
PS By the way you’ve <strike>almost</strike> convinced me on the = side.

**Comment by xanthe-cat** [comment]
Oh, one thing I forgot: I should rerun my `quick.sh` script. I’ll do that next.

I should also have said for completeness, but it seems perfectly obvious to me, that if we don’t want to go with the GPL v3 licence we can always back out before committing these changes. I think it’s sensible to choose it (given that the focus of the CC licences isn’t particularly software, unlike the GPL).

(Incidentally, to return to the question of Mersenne numbers for a moment and whether composite exponents are part of the project, it is fairly easy to modify the code to allow composite exponents to be tested, but it really opens up a can of worms – is the exponent divisible by the Fermat-PRP base? Is it going to generate garbage output on the GIGO principle? These were nice [big Gippsland worms](https://en.wikipedia.org/wiki/Giant_Gippsland_earthworm) to ponder letting out, and they are large – a bucket-sized can would probably not be enough to contain them.)

File: cofact.c [hunk 1]
```diff
@@ -981,6 +988,40 @@ int main (int argc, char **argv) {
             exit (1);
```
Comment (tdulcet) [inline cofact.c:998]: I believe you are missing a `+ 1` here in the Fermat case (line 984), as the plus one in the Fermat numbers format requires an extra byte. In other words, I believe it should be `res_len = (1 << (n_proof - 3)) + 1;`, which is consistent with [what Mlucas does](https://github.com/primesearch/Mlucas/blob/4db9fbc42ba00298c123623549bd36dd47573d45/src/Mlucas.c#L5199). (This was the only notable difference I noticed when comparing your implementation of the res2048 recovery to my GIMPS status script.)

**Comment by gary641** [comment]
I added a LICENSE file to the Readme_fixes branch using the "scales" like you suggested (LICENSE definately seems like a better name than COPYING). I picked the GNU GPL v3 without any changes, so it should be identical to the one on your branch. I also added the license notices and copyright notice to the top of my cofact.c (just a copy/paste from yours). 

Regarding composite Mersenne exponents, I think having cofact handle only prime exponents is perfectly fine, and consistent with the focus of GIMPS. No need to open up a big can of worms for no value!

I looked over your latest edits to README.md and they all look good! After you address Teal's comment I think you are ready to merge. It does appear that your merge will go into the cxc branch, as desired.

File: cofact.c (hunk 1 above)
Comment (xanthe-cat) [reply cofact.c:998]: Technically you are correct that an extra byte is required, in the case that the top most bit is not zero; however that would be the highly unusual case that the $A$ residue $\equiv -1$ (mod $F$). If $F$ is prime it should be the previous iteration’s Pépin residue that should be $-1$, which when squared results in $A \equiv 1$ (mod $F$). I will go out on a limb and suggest every residue likely to ever be imported will have the top most bit equal to zero, so the extra byte is never in practice required.

**Comment by xanthe-cat** [comment]
I already had thought over the possibility of a flag for skipping the additional proof validation step, so it was a fairly quick feature to add and test, and it also means if the full Fermat-PRP test is being calculated with -c, and the computed A value matches the A_proof value from the residue, we can avoid the additional computation for validating a proof which at the very least has the correct A value.
Basically the combination of the number being Mersenne; -j flag to add reporting; -u to use a proof; and the -e to exclude the proof from being validated, will result in the error-code field of the manual result being 0x1 instead of 0x0; that is enough for me to say this result might be not as trustworthy, but it should be reliable.
